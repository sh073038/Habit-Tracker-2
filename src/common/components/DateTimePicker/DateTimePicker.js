// @flow

import React from 'react'
import moment from 'moment/moment'
import TextField from 'material-ui/TextField'

type Props = {
  id: string,
  className?: string,
  disabled?: boolean,
  label: string,
  onChange: (?number) => void,
  value: number,
  minValue?: ?number,
}

export const DATE_TIME_FORMAT = 'YYYY-MM-DDTHH:mm'
const DATE_FORMAT = 'YYYY-MM-DD'
const TIME_FORMAT = 'HH:mm'

class DateTimePicker extends React.Component<Props> {
  handleChangeDate = (ev: any) => {
    const { value, minValue, onChange } = this.props

    if (!ev.target.value) {
      onChange(null)
      return
    }

    const dateTime: Array<string> = moment(value)
      .format(DATE_TIME_FORMAT)
      .split('T')

    if (minValue && moment(value).isBefore(moment(minValue))) {
      alert('invalid date, yo')
      return
    }

    const newValue: string = `${ev.target.value}T${dateTime[1]}`

    onChange(moment(newValue, DATE_TIME_FORMAT).valueOf())
  }

  handleChangeTime = (ev: any) => {
    const { value, minValue, onChange } = this.props

    if (!ev.target.value) {
      onChange(null)
      return
    }

    const dateTime: Array<string> = moment(value)
      .format(DATE_TIME_FORMAT)
      .split('T')

    if (minValue && moment(value).isBefore(moment(minValue))) {
      alert('invalid date, yo')
      return
    }

    const newValue: string = `${dateTime[0]}T${ev.target.value}`

    onChange(moment(newValue, DATE_TIME_FORMAT).valueOf())
  }

  render() {
    const { id, label, value, className = '', disabled = false } = this.props

    return (
      <div className={className}>
        <TextField
          id={`date-${id}`}
          label={label}
          type="date"
          value={moment(value).format(DATE_FORMAT)}
          InputLabelProps={{
            shrink: true,
          }}
          disabled={disabled}
          onChange={this.handleChangeDate}
        />
        <TextField
          id={`time-${id}`}
          type="time"
          value={moment(value).format(TIME_FORMAT)}
          InputLabelProps={{
            shrink: true,
          }}
          inputProps={{
            step: 600, // 10 min
          }}
          disabled={disabled}
          onChange={this.handleChangeTime}
        />
      </div>
    )
  }
}

export default DateTimePicker
