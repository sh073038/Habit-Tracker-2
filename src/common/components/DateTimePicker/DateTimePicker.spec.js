import React from 'react'
import { shallow } from 'enzyme'
import moment from 'moment'

import DateTimePicker from './DateTimePicker'

describe('DateTimePicker', () => {
  const dateNow = Date.now

  beforeEach(() => {
    Date.now = jest.fn(() => 1518110488752) // lock at "2018-02-08T18:21:45+01:00"
  })

  afterEach(() => {
    Date.now = dateNow
  })

  it('renders component', () => {
    const wrapper = shallow(
      <DateTimePicker
        id="test-picker"
        label="Date Time"
        value={moment()
          .subtract(5, 'd')
          .valueOf()}
        onChange={null}
      />,
    )
    expect(wrapper.getElement()).toMatchSnapshot()
    expect(wrapper.props()).toMatchSnapshot()
  })

  it('value change work', () => {
    const val1 = moment()
      .subtract(5, 'd')
      .valueOf()

    const val2 = moment()
      .add(5, 'd')
      .valueOf()

    let wrapper = shallow(
      <DateTimePicker id="test-picker" label="Date Time" value={val1} onChange={null} />,
    )
    expect(
      wrapper
        .children()
        .first()
        .props().value,
    ).toEqual('2018-02-03')
    expect(
      wrapper
        .children()
        .last()
        .props().value,
    ).toEqual('18:21')
    wrapper = shallow(
      <DateTimePicker id="test-picker" label="Date Time" value={val2} onChange={null} />,
    )
    expect(
      wrapper
        .children()
        .first()
        .props().value,
    ).toEqual('2018-02-13')
    expect(
      wrapper
        .children()
        .last()
        .props().value,
    ).toEqual('18:21')
  })

  it('works with minValue', () => {
    // TODO
  })
})
