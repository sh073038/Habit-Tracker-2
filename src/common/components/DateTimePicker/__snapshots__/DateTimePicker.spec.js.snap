// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DateTimePicker renders component 1`] = `
<div
  className=""
>
  <TextField
    InputLabelProps={
      Object {
        "shrink": true,
      }
    }
    disabled={false}
    id="date-test-picker"
    label="Date Time"
    onChange={[Function]}
    required={false}
    select={false}
    type="date"
    value="2018-02-03"
  />
  <TextField
    InputLabelProps={
      Object {
        "shrink": true,
      }
    }
    disabled={false}
    id="time-test-picker"
    inputProps={
      Object {
        "step": 600,
      }
    }
    onChange={[Function]}
    required={false}
    select={false}
    type="time"
    value="18:21"
  />
</div>
`;

exports[`DateTimePicker renders component 2`] = `
Object {
  "children": Array [
    <TextField
      InputLabelProps={
        Object {
          "shrink": true,
        }
      }
      disabled={false}
      id="date-test-picker"
      label="Date Time"
      onChange={[Function]}
      required={false}
      select={false}
      type="date"
      value="2018-02-03"
    />,
    <TextField
      InputLabelProps={
        Object {
          "shrink": true,
        }
      }
      disabled={false}
      id="time-test-picker"
      inputProps={
        Object {
          "step": 600,
        }
      }
      onChange={[Function]}
      required={false}
      select={false}
      type="time"
      value="18:21"
    />,
  ],
  "className": "",
}
`;
